server/sync.go

```diff
@@ -0,0 +1,223 @@
+					continue
+				}
+
+				max := time.UnixMilli(meta.MaxTime)
+				if max.After(ret) {
+					blocksToSync = append(blocksToSync, entry.Name())
+				}
+			}
+
+			_, err = p.fileBackend.FileExists(PluginName)
+			if err != nil {
+				p.API.LogError("unable check file store, exiting sync loop", "err", err)
+				return
+			}
+
+			// we are using a pre-defined directory structure in the filestore
+			// hence using the same remoteStorageDir variable here
+			blocksInSFileStore := make(map[string]bool)
+			dirs, err2 := p.fileBackend.ListDirectory(remoteStorageDir)
+			if err2 != nil {
+				p.API.LogError("could not list directory from filestore", "err", err2)
+			}
+			for i := range dirs {
+				// we trim the parent dir from the dirs as filebackend returns the
+				// absolute path (relative to filestore root)
+				dir := strings.TrimPrefix(dirs[i], remoteStorageDir+string(os.PathSeparator))
+				blocksInSFileStore[dir] = true
+			}
+
+			// we should skip the blocks those are already in the filestore
+			// since the tsdb blocks are immutable, we don't need to worry whether
+			// if the block is changed over time
+			blocksToSync = slices.DeleteFunc(blocksToSync, func(s string) bool {
+				_, ok := blocksInSFileStore[s]
+				return ok
+			})
+
+			for _, block := range blocksToSync {
+				err2 := copyDirectory(
+					filepath.Join(localStorageDir, block),
+					filepath.Join(remoteStorageDir, block),
+					p.fileBackend.WriteFile,
+				)
+				if err2 != nil {
+					p.API.LogError("could not write block to filestore", "err", err2)
+				}
+			}
+		case <-tickFileStoreCleanUp:
+			p.API.LogDebug("Cleaning up the filestore...")
+			ret := time.Now().AddDate(0, 0, -1**p.configuration.RetentionDurationDays)
+
+			// get the blocks if there is any block in the remote filestore
+			blocks, err := p.fileBackend.ListDirectory(remoteStorageDir)
+			if err != nil {
+				p.API.LogError("unable check file store, skipping filestore cleanup", "err", err)
+				continue
+			}
+
+			// read block meta from the remote filestore and decide if they are older than the
+			// retention period. If so, delete.
+			for _, b := range blocks {
+				meta, err := readBlockMeta(filepath.Join(b, metaFileName), p.fileBackend.ReadFile)
+				if err != nil {
+					// we intentionally log with debug level here, file store returns wrapped errors
+					// and to not pollute the logs, we simply reducing the log level here
+					p.API.LogDebug("unable to read meta file", "err", err)
+					continue
+				}
+
+				max := time.UnixMilli(meta.MaxTime)
+				if max.Before(ret) {
+					p.API.LogInfo("Deleting obsolete block from the filestore", "ulid", meta.ULID, "Max Time", max.String())
+					err = p.fileBackend.RemoveDirectory(b)
+					if err != nil {
+						p.API.LogWarn("unable to remove block from filestore", "err", err)
+					}
+				}
+			}
+
+			p.API.LogDebug("Cleaned up the filestore")
+		case <-p.closeChan:
+			p.API.LogDebug("Filestore sync job stopped")
```

Comment:
On shutdown, we should do one final sync to push out any unsynced files to the remote store. This is to guard against cases in cloud env where the local filestore is ephemeral. So during pod rollover, we would lose out on any unsynced data.

Reply:
Makes sense, I guess if we use `time.NewTicker`, it can be handled by closing the ticker right?

Reply:
Yes, it can be done in the defer clause as well.

Reply:
Gotcha 👍